// ngxtension/esm2022/trackby-id-prop/trackby-id-prop.mjs
//INSPIRED BY https://medium.com/ngconf/make-trackby-easy-to-use-a3dd5f1f733b
import { NgForOf } from '@angular/common';
import { Directive, Input, inject } from '@angular/core';
import * as i0 from "@angular/core";
export class TrackById {
    constructor() {
        this.ngFor = inject((NgForOf), { self: true });
        this.ngFor.ngForTrackBy = (index, item) => item.id;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: TrackById, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: TrackById, isStandalone: true, selector: "[ngForTrackById]", inputs: { ngForOf: "ngForOf" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: TrackById, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngForTrackById]',
                    standalone: true,
                }]
        }], ctorParameters: () => [], propDecorators: { ngForOf: [{
                type: Input
            }] } });
export class TrackByProp {
    constructor() {
        this.ngFor = inject((NgForOf), { self: true });
    }
    set ngForTrackByProp(trackByProp) {
        if (!trackByProp)
            return;
        this.ngFor.ngForTrackBy = (index, item) => item[trackByProp];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: TrackByProp, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: TrackByProp, isStandalone: true, selector: "[ngForTrackByProp]", inputs: { ngForOf: "ngForOf", ngForTrackByProp: "ngForTrackByProp" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: TrackByProp, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngForTrackByProp]',
                    standalone: true,
                }]
        }], propDecorators: { ngForOf: [{
                type: Input
            }], ngForTrackByProp: [{
                type: Input,
                args: [{ required: true }]
            }] } });
export const TRACK_BY_DIRECTIVES = [TrackById, TrackByProp];
