// ngxtension/esm2022/gestures/drag.mjs
import { Directive, EventEmitter, Injector, Input, Output, inject, signal, } from '@angular/core';
import { DragGesture } from '@use-gesture/vanilla';
import { createGesture } from './gesture';
import * as i0 from "@angular/core";
export const injectDrag = createGesture('drag', DragGesture);
export class NgxDrag {
    constructor() {
        this.config = signal({});
        this.ngxDrag = new EventEmitter();
        this.injector = inject(Injector);
    }
    set _config(config) {
        this.config.set(config);
    }
    ngOnInit() {
        injectDrag(this.ngxDrag.emit.bind(this.ngxDrag), {
            injector: this.injector,
            zoneless: this.zoneless,
            config: this.config,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxDrag, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: NgxDrag, isStandalone: true, selector: "[ngxDrag]", inputs: { _config: ["ngxDragConfig", "_config"], zoneless: ["ngxDragZoneless", "zoneless"] }, outputs: { ngxDrag: "ngxDrag" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxDrag, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxDrag]',
                    standalone: true,
                }]
        }], propDecorators: { _config: [{
                type: Input,
                args: ['ngxDragConfig']
            }], zoneless: [{
                type: Input,
                args: ['ngxDragZoneless']
            }], ngxDrag: [{
                type: Output
            }] } });
