// ngxtension/esm2022/gestures/hover.mjs
import { Directive, EventEmitter, Injector, Input, Output, inject, signal, } from '@angular/core';
import { HoverGesture } from '@use-gesture/vanilla';
import { createGesture } from './gesture';
import * as i0 from "@angular/core";
export const injectHover = createGesture('hover', HoverGesture);
export class NgxHover {
    constructor() {
        this.config = signal({});
        this.ngxHover = new EventEmitter();
        this.injector = inject(Injector);
    }
    set _config(config) {
        this.config.set(config);
    }
    ngOnInit() {
        injectHover(this.ngxHover.emit.bind(this.ngxHover), {
            injector: this.injector,
            zoneless: this.zoneless,
            config: this.config,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxHover, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: NgxHover, isStandalone: true, selector: "[ngxHover]", inputs: { _config: ["ngxHoverConfig", "_config"], zoneless: ["ngxHoverZoneless", "zoneless"] }, outputs: { ngxHover: "ngxHover" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxHover, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxHover]',
                    standalone: true,
                }]
        }], propDecorators: { _config: [{
                type: Input,
                args: ['ngxHoverConfig']
            }], zoneless: [{
                type: Input,
                args: ['ngxHoverZoneless']
            }], ngxHover: [{
                type: Output
            }] } });
