// ngxtension/esm2022/gestures/move.mjs
import { Directive, EventEmitter, Injector, Input, Output, inject, signal, } from '@angular/core';
import { MoveGesture } from '@use-gesture/vanilla';
import { createGesture } from './gesture';
import * as i0 from "@angular/core";
export const injectMove = createGesture('move', MoveGesture);
export class NgxMove {
    constructor() {
        this.config = signal({});
        this.ngxMove = new EventEmitter();
        this.injector = inject(Injector);
    }
    set _config(config) {
        this.config.set(config);
    }
    ngOnInit() {
        injectMove(this.ngxMove.emit.bind(this.ngxMove), {
            injector: this.injector,
            zoneless: this.zoneless,
            config: this.config,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxMove, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: NgxMove, isStandalone: true, selector: "[ngxMove]", inputs: { _config: ["ngxMoveConfig", "_config"], zoneless: ["ngxMoveZoneless", "zoneless"] }, outputs: { ngxMove: "ngxMove" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxMove, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxMove]',
                    standalone: true,
                }]
        }], propDecorators: { _config: [{
                type: Input,
                args: ['ngxMoveConfig']
            }], zoneless: [{
                type: Input,
                args: ['ngxMoveZoneless']
            }], ngxMove: [{
                type: Output
            }] } });
