// ngxtension/esm2022/gestures/pinch.mjs
import { Directive, EventEmitter, Injector, Input, Output, inject, signal, } from '@angular/core';
import { PinchGesture } from '@use-gesture/vanilla';
import { createGesture } from './gesture';
import * as i0 from "@angular/core";
export const injectPinch = createGesture('pinch', PinchGesture);
export class NgxPinch {
    constructor() {
        this.config = signal({});
        this.ngxPinch = new EventEmitter();
        this.injector = inject(Injector);
    }
    set _config(config) {
        this.config.set(config);
    }
    ngOnInit() {
        injectPinch(this.ngxPinch.emit.bind(this.ngxPinch), {
            injector: this.injector,
            zoneless: this.zoneless,
            config: this.config,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxPinch, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: NgxPinch, isStandalone: true, selector: "[ngxPinch]", inputs: { _config: ["ngxPinchConfig", "_config"], zoneless: ["ngxPinchZoneless", "zoneless"] }, outputs: { ngxPinch: "ngxPinch" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxPinch, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxPinch]',
                    standalone: true,
                }]
        }], propDecorators: { _config: [{
                type: Input,
                args: ['ngxPinchConfig']
            }], zoneless: [{
                type: Input,
                args: ['ngxPinchZoneless']
            }], ngxPinch: [{
                type: Output
            }] } });
