// ngxtension/esm2022/gestures/scroll.mjs
import { Directive, EventEmitter, Injector, Input, Output, inject, signal, } from '@angular/core';
import { ScrollGesture } from '@use-gesture/vanilla';
import { createGesture } from './gesture';
import * as i0 from "@angular/core";
export const injectScroll = createGesture('scroll', ScrollGesture);
export class NgxScroll {
    constructor() {
        this.config = signal({});
        this.ngxScroll = new EventEmitter();
        this.injector = inject(Injector);
    }
    set _config(config) {
        this.config.set(config);
    }
    ngOnInit() {
        injectScroll(this.ngxScroll.emit.bind(this.ngxScroll), {
            injector: this.injector,
            zoneless: this.zoneless,
            config: this.config,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxScroll, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: NgxScroll, isStandalone: true, selector: "[ngxScroll]", inputs: { _config: ["ngxScrollConfig", "_config"], zoneless: ["ngxScrollZoneless", "zoneless"] }, outputs: { ngxScroll: "ngxScroll" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxScroll, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxScroll]',
                    standalone: true,
                }]
        }], propDecorators: { _config: [{
                type: Input,
                args: ['ngxScrollConfig']
            }], zoneless: [{
                type: Input,
                args: ['ngxScrollZoneless']
            }], ngxScroll: [{
                type: Output
            }] } });
