// ngxtension/esm2022/gestures/wheel.mjs
import { Directive, EventEmitter, Injector, Input, Output, inject, signal, } from '@angular/core';
import { WheelGesture } from '@use-gesture/vanilla';
import { createGesture } from './gesture';
import * as i0 from "@angular/core";
export const injectWheel = createGesture('wheel', WheelGesture);
export class NgxWheel {
    constructor() {
        this.config = signal({});
        this.ngxWheel = new EventEmitter();
        this.injector = inject(Injector);
    }
    set _config(config) {
        this.config.set(config);
    }
    ngOnInit() {
        injectWheel(this.ngxWheel.emit.bind(this.ngxWheel), {
            injector: this.injector,
            zoneless: this.zoneless,
            config: this.config,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxWheel, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.3", type: NgxWheel, isStandalone: true, selector: "[ngxWheel]", inputs: { _config: ["ngxWheelConfig", "_config"], zoneless: ["ngxWheelZoneless", "zoneless"] }, outputs: { ngxWheel: "ngxWheel" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: NgxWheel, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxWheel]',
                    standalone: true,
                }]
        }], propDecorators: { _config: [{
                type: Input,
                args: ['ngxWheelConfig']
            }], zoneless: [{
                type: Input,
                args: ['ngxWheelZoneless']
            }], ngxWheel: [{
                type: Output
            }] } });
