// ngxtension/esm2022/intl/display-names.pipe.mjs
import { inject, LOCALE_ID, Pipe, } from '@angular/core';
import { createInjectionToken } from 'ngxtension/create-injection-token';
import * as i0 from "@angular/core";
/**
 * @internal
 */
const defaultOptions = {
    style: 'short',
    localeMatcher: 'lookup',
    fallback: 'code',
};
/**
 * @internal
 */
const [injectFn, provideFn] = createInjectionToken(() => defaultOptions);
/**
 * Provides a way to inject the options for the DisplayNamesPipe.
 *
 * @param options The options to use for the DisplayNamesPipe.
 * @returns The provider for the DisplayNamesPipe.
 */
export function provideDisplayNamesOptions(options) {
    return provideFn({ ...defaultOptions, ...options });
}
/**
 * This pipe is a wrapper around the [Intl.DisplayNames](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames) API.
 *
 * @returns The display name of the code or the code as it is in case of errors.
 */
export class DisplayNamesPipe {
    constructor() {
        this.defaultOptions = injectFn();
        this.locale = inject(LOCALE_ID);
    }
    /**
     * Displays the name of the given code in the given locale.
     *
     * @param code The code to transform.
     * @param type DisplayNamesType to use.
     * @param style Optional. The formatting style to use. Defaults to "short".
     * @param locale Optional. The locale to use for the transformation. Defaults to LOCALE_ID.
     * @returns The name of the given code in the given locale or the code itself if the name could not be found.
     */
    transform(code, type, style, locale) {
        try {
            return new Intl.DisplayNames(locale || this.locale, {
                ...this.defaultOptions,
                type,
                ...(style ? { style } : {}),
            }).of(code);
        }
        catch (e) {
            console.error(e);
            return code;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: DisplayNamesPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.3", ngImport: i0, type: DisplayNamesPipe, isStandalone: true, name: "displayNames" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: DisplayNamesPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'displayNames',
                    standalone: true,
                }]
        }] });
