// ngxtension/esm2022/intl/list-format.pipe.mjs
import { inject, LOCALE_ID, Pipe, } from '@angular/core';
import { createInjectionToken } from 'ngxtension/create-injection-token';
import * as i0 from "@angular/core";
/**
 * @internal
 */
const defaultOptions = {
    style: 'long',
    type: 'conjunction',
};
/**
 * @internal
 */
const [injectFn, provideFn] = createInjectionToken(() => defaultOptions);
/**
 * Provides a way to inject the options for the ListFormatPipe.
 *
 * @param options The options to use for the ListFormatPipe.
 * @returns The provider for the ListFormatPipe.
 */
export function provideListFormatOptions(options) {
    return provideFn({ ...defaultOptions, ...options });
}
/**
 * This pipe is a wrapper around the [Intl.ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat) API.
 *
 * @returns The formatted list of values or the list as string in case of errors.
 */
export class ListFormatPipe {
    constructor() {
        this.defaultOptions = injectFn();
        this.locale = inject(LOCALE_ID);
    }
    /**
     * Transforms the list of values into a formatted string.
     *
     * @param value The list of values to format.
     * @param style Optional. The formatting style to use. Defaults to "long".
     * @param locale Optional. The locale to use for the transformation. Defaults to LOCALE_ID.
     * @returns The formatted list of values or the list as string in case of errors.
     */
    transform(value, style, locale) {
        try {
            return new Intl.ListFormat(locale || this.locale, {
                ...this.defaultOptions,
                ...(style ? { style } : {}),
            }).format(Array.from(value));
        }
        catch (e) {
            console.error(e);
            return Array.from(value).join(', ');
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: ListFormatPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.3", ngImport: i0, type: ListFormatPipe, isStandalone: true, name: "listFormat" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: ListFormatPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'listFormat',
                    standalone: true,
                }]
        }] });
