// ngxtension/esm2022/intl/plural-rules.pipe.mjs
import { inject, LOCALE_ID, Pipe, } from '@angular/core';
import { createInjectionToken } from 'ngxtension/create-injection-token';
import * as i0 from "@angular/core";
/**
 * @internal
 */
const defaultOptions = {
    localeMatcher: 'best fit', // other values: "lookup",
    type: 'cardinal', // other values: "ordinal"
};
/**
 * @internal
 */
const [injectFn, provideFn] = createInjectionToken(() => defaultOptions);
/**
 * Provides a way to inject the options for the PluralRules.
 *
 * @param options The options to use for the PluralRules.
 * @returns The provider for the PluralRules.
 */
export function providePluralRulesOptions(options) {
    return provideFn({ ...defaultOptions, ...options });
}
/**
 * This pipe is a wrapper around the [Intl.PluralRules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules) API.
 * It takes a value and returns the plural category for that value.
 *
 * @returns The plural category for the value or the value as string in case of errors.
 */
export class PluralRulesPipe {
    constructor() {
        this.defaultOptions = injectFn();
        this.locale = inject(LOCALE_ID);
    }
    /**
     * Transforms the value into a plural category.
     *
     * @param value The value to transform.
     * @param locale Optional, the locale to use for the formatting.
     * @returns The plural category for the value or the value as string in case of errors.
     */
    transform(value, locale) {
        try {
            return new Intl.PluralRules(locale || this.locale, this.defaultOptions).select(value);
        }
        catch (e) {
            console.error(e);
            return String(value);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PluralRulesPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.3", ngImport: i0, type: PluralRulesPipe, isStandalone: true, name: "pluralRules" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: PluralRulesPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'pluralRules',
                    standalone: true,
                }]
        }] });
