// ngxtension/esm2022/intl/relative-time-format.pipe.mjs
import { inject, LOCALE_ID, Pipe, } from '@angular/core';
import { createInjectionToken } from 'ngxtension/create-injection-token';
import * as i0 from "@angular/core";
/**
 * @internal
 */
const defaultOptions = {
    localeMatcher: 'best fit', // other values: "lookup"
    numeric: 'always', // other values: "auto"
    style: 'long', // other values: "short" or "narrow"
};
/**
 * @internal
 */
const [injectFn, provideFn] = createInjectionToken(() => defaultOptions);
/**
 * Provides a way to inject the options for the RelativeTimeFormatPipe.
 * @param options The options to use for the RelativeTimeFormatPipe.
 *
 * @returns The provider for the RelativeTimeFormatPipe.
 */
export function provideRelativeTimeFormatOptions(options) {
    return provideFn({ ...defaultOptions, ...options });
}
/**
 * This pipe is a wrapper around the [Intl.RelativeTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat) API.
 *
 * @returns The relative time format of the value or the value as it is in case of errors.
 */
export class RelativeTimeFormatPipe {
    constructor() {
        this.defaultOptions = injectFn();
        this.locale = inject(LOCALE_ID);
    }
    /**
     * Transforms the value into a relative time format.
     *
     * @param value The value to format.
     * @param unit The unit of the value.
     * @param style Optional, the formatting style to use.
     * @param locale Optional, the locale to use for the formatting.
     * @returns The relative time format of the value or the value as it is in case of errors.
     */
    transform(value, unit, style, locale) {
        try {
            return new Intl.RelativeTimeFormat(locale || this.locale, {
                ...this.defaultOptions,
                ...(style ? { style } : {}),
            }).format(value, unit);
        }
        catch (e) {
            console.error(e);
            return value.toString();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: RelativeTimeFormatPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.3", ngImport: i0, type: RelativeTimeFormatPipe, isStandalone: true, name: "relativeTimeFormat" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: RelativeTimeFormatPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'relativeTimeFormat',
                    standalone: true,
                }]
        }] });
