// ngxtension/esm2022/intl/supportedValuesOf.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * This pipe is a wrapper around the [Intl.supportedValuesOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/supportedValuesOf) method.
 *
 * @returns An array containing the supported calendar, collation, currency, numbering systems, or unit values supported by the implementation.
 */
export class SupportedValuesOf {
    /**
     * Transforms a key into an array containing the supported calendar, collation, currency, numbering systems, or unit values supported by the implementation.
     *
     * @param key A key string indicating the category of values to be returned. This is one of: "calendar", "collation", "currency","numberingSystem", "timeZone", "unit"
     * @returns An array containing the supported values for the key or an empty array if the input is invalid.
     */
    transform(key) {
        try {
            return Intl.supportedValuesOf(key);
        }
        catch (e) {
            console.error(e);
            return [];
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: SupportedValuesOf, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.3", ngImport: i0, type: SupportedValuesOf, isStandalone: true, name: "supportedValuesOf" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.3", ngImport: i0, type: SupportedValuesOf, decorators: [{
            type: Pipe,
            args: [{
                    name: 'supportedValuesOf',
                    standalone: true,
                }]
        }] });
